src/hnsw/build.c

```diff
@@ -132,7 +134,59 @@ static void BuildCallback(
     MemoryContextReset(buildstate->tmpCtx);
 }
 
-static int GetArrayLengthFromHeap(Relation heap, int indexCol)
+static int GetArrayLengthFromExpression(Expr *expression, Relation heap, HeapTuple tuple)
+{
+    ExprContext    *econtext;
+    ExprState      *exprstate;
+    EState         *estate;
+    Datum           result;
+    bool            isNull;
+    ArrayType      *array;
+    int             n_items = HNSW_DEFAULT_DIM;
+    TupleTableSlot *slot;
+    TupleDesc       tupdesc = RelationGetDescr(heap);
+
+#if PG_VERSION_NUM >= 120000
+    slot = MakeSingleTupleTableSlot(tupdesc, &TTSOpsHeapTuple);
+#else
+    slot = MakeSingleTupleTableSlot(tupdesc);
+#endif
+
+    // Create an expression context
+    econtext = CreateStandaloneExprContext();
+    estate = CreateExecutorState();
+
+    // Build the expression state for your expression
+    exprstate = ExecPrepareExpr(expression, estate);
+
+#if PG_VERSION_NUM >= 120000
+    ExecStoreHeapTuple(tuple, slot, false);
+#else
+    ExecStoreTuple(tuple, slot, InvalidBuffer, false);
+#endif
+    // Set up the tuple for the expression evaluation
+    econtext->ecxt_scantuple = slot;
+
+    // Evaluate the expression for the first row
+    result = ExecEvalExprSwitchContext(exprstate, econtext, &isNull);
+
+    // Release tuple descriptor
+    ReleaseTupleDesc(tupdesc);
+
+    // Check if the result is not null
+    if(!isNull) {
+        // todo check if datum is array
+        array = DatumGetArrayTypePCopy(result);
+        n_items = ArrayGetNItems(ARR_NDIM(array), ARR_DIMS(array));
+        return n_items;
+    } else {
+        elog(ERROR, "Expression did not result in an array");
```

Comment:
Maybe give more context in the error?
e.g.
`Expression used in CREATE INDEX statement did not result in hnsw-index compatible array`

Keep in mind that this is all the user will see when they run an illegal index creation like
`CREATE INDEX ON papers(TO_UPPERCASE(abstract));`. So the error should give them a clue what went wrong.